tools/autograd/gen_variable_type.py

```diff
@@ -462,6 +468,21 @@ def emit_increment_version():
     combined = nested_dict(env, declaration)
 
     body = []
+    if declaration['zero_dim_dispatch_from']:
+        # We can't use the scalar code path for scalars that are backed by tensors, because
+        # it would break autodiff wrt those scalars. That's why we reshape them to be 1d
+        # and dispatch again.
+        # XXX: using self as a template for the valid type for the arg is a bit aribitrary,
+        # but seems to work well. We don't have any concrete type labels in declarations, so
+        # we can't do anything better for now.
+        # XXX: scalars and 1-element tensors are equivalent in terms of broadcasting, but
```

Comment:
can you show examples of the generated code changing, both in function_wrapper generated files and VariableType?  This makes me pretty nervous, because e.g. the above statement isn't true; scalars can be broadcast to 1-element tensors but the reverse isn't true.  This usually only comes up when you do an in-place op, where the broadcasting only goes in one direction.